### PR TITLE
Introduce new class: Pipenv

### PR DESCRIPTION
**Pull request by nre-ableton**
This class is designed to facilitate running a Pipenv command against
multiple versions of Python, which is something that Pipenv cannot do
by itself very gracefully. We do this by using the `--python` argument
to Pipenv, which assumes that the `Pipfile` does not contain a
`requirements` section which may cause Pipenv to throw an error.

---

This is the first step in solving https://github.com/AbletonAppDev/devtools/issues/1340, ping @AbletonDevTools/gotham-city

### CODE REVIEW

**Comment by nre-ableton** [comment]
re-ping @AbletonDevTools/gotham-city

File: src/com/ableton/Pipenv.groovy [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    assert script
+    assert args
+    assert args.command
+    assert args.pythonVersions
+
+    Map result = [:]
+
+    try {
+      args.pythonVersions.each { python ->
+        script.sh "pipenv install --dev --python ${python}"
```
Comment (ala-ableton) [inline src/com/ableton/Pipenv.groovy]: It should be documented somewhere that we are passing `--dev` to `pipenv install`.

File: src/com/ableton/Pipenv.groovy [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+   *                 <strong>Note:</strong> this option and the {@code returnStatus}
+   *                 option are mutually exclusive.
+   *               </li>
+   *             </ul>
+   * @return Map of return values. The keys in the map correspond to the Python versions
+   *         given in {@code args.pythonVersions}, and the values are either integers,
+   *         strings, or null, depending on whether {@code returnStatus},
+   *         {@code returnStdout}, or neither option was specified (respectively).
+   */
+  Map runCommand(Map args) {
```
Comment (ala-ableton) [inline src/com/ableton/Pipenv.groovy]: Why do we need both `runCommand` and `runWith`? `runCommand` seems to be a subset of `runWith`.

File: src/com/ableton/Pipenv.groovy [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+  Map runWith(Map args, Closure body) {
+    assert script
+    assert args
+    assert args.pythonVersions
+
+    Map result = [:]
+
+    try {
+      args.pythonVersions.each { python ->
+        script.sh "pipenv install --dev --python ${python}"
```
Comment (ala-ableton) [inline src/com/ableton/Pipenv.groovy]: It should be documented somewhere that we are passing `--dev` to `pipenv install`.

File: src/com/ableton/Pipenv.groovy (hunk 2 above)
Comment (nre-ableton) [reply src/com/ableton/Pipenv.groovy]: It's a convenience, but I'm open to removing `runCommand` and only having `runWith`.

File: src/com/ableton/Pipenv.groovy (hunk 3 above)
Comment (nre-ableton) [reply src/com/ableton/Pipenv.groovy]: Fixed

File: src/com/ableton/Pipenv.groovy (hunk 1 above)
Comment (nre-ableton) [reply src/com/ableton/Pipenv.groovy]: I've removed this method (see below)

File: vars/pipenv.groovy [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+
+/**
+ * Run a command with Pipenv using multiple versions of Python.
+ * @param args Map of arguments.
+ * @return Map with return output or values for each Python version.
+ * @see com.ableton.Pipenv#runCommand(Map)
+ */
+Map runCommand(Map args) {
+  return new Pipenv(script: this).runCommand(args)
```
Comment (ala-ableton) [inline vars/pipenv.groovy]: @nre-ableton this should be removed.

Comment (nre-ableton) [reply vars/pipenv.groovy]: @ala-ableton Good catch! Removed.

File: src/com/ableton/Pipenv.groovy [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+   * parallelized and therefore the commands are run serially for each Python version.
+   *
+   * This function also installs the development packages for each Python version (in
+   * other words, it runs {@code pipenv install --dev}. Also, it removes the virtualenv
+   * after the last Python version has been run.
+   *
+   * @param pythonVersions List of Python versions to run the command with. This argument
+   *                       is passed to Pipenv via {@code pipenv --python}. See
+   *                       {@code pipenv --help} for supported syntax.
+   * @param body Closure body to execute. The closure body is passed the python version
```
Comment (ala-ableton) [inline src/com/ableton/Pipenv.groovy]: Nitpick: `python version` -> `Python version`

File: src/com/ableton/Pipenv.groovy [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+   *             as an argument.
+   * @return Map of return values. The keys in the map correspond to the Python versions
+   *         given in {@code args.pythonVersions}, and the values are the result of
```
Comment (ala-ableton) [inline src/com/ableton/Pipenv.groovy]: `are the results`?

File: src/com/ableton/Pipenv.groovy [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+   *         executing the closure body.
+   */
+  Map runWith(List pythonVersions, Closure body) {
+    assert script
+    assert pythonVersions
+
+    Map result = [:]
```
Comment (ala-ableton) [inline src/com/ableton/Pipenv.groovy]: `result` -> `results`?

File: src/com/ableton/Pipenv.groovy (hunk 6 above)
Comment (nre-ableton) [reply src/com/ableton/Pipenv.groovy]: Fixed, and also improved the surrounding grammar.

**Comment by nre-ableton** [comment]
@ala-ableton Regarding the green PR on another repo: https://github.com/AbletonDevTools/ablbot/pull/186

I'm going to go ahead and merge this change, assuming there are no objections.

**Pull request merged**